src/mame/drivers/gi6809.cpp

```diff
@@ -0,0 +1,1039 @@
+	required_device<screen_device> m_screen;
+	required_device<gfxdecode_device> m_gfxdecode;
+	required_device<discrete_device> m_discrete;
+	required_shared_ptr<uint8_t> m_videoram;
+	required_shared_ptr<uint8_t> m_colorram;
+	required_ioport_array<4> m_input;
+	optional_ioport m_coin_gl;
+	optional_ioport m_coin_ca;
+	optional_device<ticket_dispenser_device> m_hopper;
+	output_finder<8> m_lamps;
+
+	// Video Hardware
+	TILE_GET_INFO_MEMBER(get_bg_tile_info);
+	tilemap_t *m_bg_tilemap;
+	void gi6809_videoram_w(offs_t offset, uint8_t data);
+	void gi6809_colorram_w(offs_t offset, uint8_t data);
+	uint32_t screen_update_gi6809(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect);
+	void gi6809_palette(palette_device &palette) const;
+	void castawayt_palette(palette_device &palette) const;
+
+	// Input ports
+	uint8_t gi6809_mux_port_r();
+	uint8_t cast_mux_port_r();
+
+	// Output ports
+	void lamps3_w(uint8_t data);
+	void lamps3h_w(uint8_t data);
+	void lamps5_w(uint8_t data);
+	void lamps8_w(uint8_t data);
+	void snd_mux_w(uint8_t data);
+
+	// Input-Output lines
+	DECLARE_READ_LINE_MEMBER(pia0_ca1);
+	DECLARE_READ_LINE_MEMBER(cast_pia1_ca1);  // castaway - coin system
+	DECLARE_READ_LINE_MEMBER(cast_pia1_ca2);  // castaway - coin system
+	DECLARE_READ_LINE_MEMBER(ret_vsync);
+
+	DECLARE_WRITE_LINE_MEMBER(crtc_vs_gl);
+	DECLARE_WRITE_LINE_MEMBER(crtc_vs_ca);
+	DECLARE_WRITE_LINE_MEMBER(crtc_vs_je);
+
+	// Latches
+	uint8_t m_mux_data = 0xff;
+	uint8_t m_cast_sense = 0xff;
+
+	// states
+	bool m_cast_ca1 = true;
+	bool m_cast_ca2 = true;
+	bool m_vsync = 0;
+
+	// internal
+	uint8_t m_stage = 0;
+	uint8_t m_ser_ptr = 0;
+	uint8_t m_sernum1[0x10] = { 3,0,3,4,4,1,4,1,0,0,0,0,0,0,2,6};
+	uint8_t m_sernum2[0x10] = { 4,7,4,9,2,0xd,4,3,0,0,0,0,0,0,0,1};
+	uint8_t sn_read1_r() { return m_sernum1[(m_ser_ptr++)& 0x0f]; }
+	uint8_t sn_read2_r() { return m_sernum2[(m_ser_ptr++)& 0x0f]; }
+	uint8_t cast_sens_r() { return m_cast_sense; }
+};
+
+
+/********************************
+*       Interrupt Control       *
+********************************/
+
+WRITE_LINE_MEMBER( gi6809_state::crtc_vs_gl )
+{
+	m_vsync = state;
+}
+
+
+WRITE_LINE_MEMBER( gi6809_state::crtc_vs_ca )
+{
+	if(state && ((m_coin_ca->read() == 0) || (m_stage != 0)))
+	{
+		switch(m_stage)
+		{
+			case 0: m_stage++;
+					m_cast_ca1 = true;
+					m_cast_ca2 = false;
+					m_cast_sense = 0xbf;  // upper coin
+					break;
+
+			case 1: m_stage++;
+					m_cast_ca1 = false;
+					m_cast_ca2 = true;
+					m_cast_sense = 0x7f;  // lower coin
+					break;
+
+			case 2: m_stage++;
+					m_cast_ca1 = true;
+					m_cast_ca2 = true;
+					m_cast_sense = 0xff;  // none
+					break;
+
+			case 3: m_stage = 0;
+					break;
+		}
+	}
+	m_vsync = state;
+}
+
+
+WRITE_LINE_MEMBER( gi6809_state::crtc_vs_je )
+{
+	m_vsync = state;
+	m_pia[1]->cb1_w(state);   // IRQB-> IRQ Master
+	m_pia[1]->ca1_w(!state);  // IRQA-> FIRQ Master
+	m_pia[0]->ca1_w(state);   // IRQA-> IRQ Slave
```

Comment:
Mixing calls to the members for setting these inputs with read callbacks is going to lead to confusing behaviour.